proposals/record-structs.md

```diff
@@ -0,0 +1,301 @@
+
+### Deconstruct
+
+A positional record struct with at least one parameter synthesizes a public void-returning instance method called Deconstruct with an out
+parameter declaration for each parameter of the primary constructor declaration. Each parameter
+of the Deconstruct method has the same type as the corresponding parameter of the primary
+constructor declaration. The body of the method assigns each parameter of the Deconstruct method
+to the value from an instance member access to a member of the same name.
+The method can be declared explicitly. It is an error if the explicit declaration does not match
+the expected signature or accessibility, or is static.
+
+# Allow `with` expression on structs
+
+It is now valid for the receiver in a `with` expression to have a struct type.
+
+On the right hand side of the `with` expression is a `member_initializer_list` with a sequence
+of assignments to *identifier*, which must be an accessible instance field or property of the receiver's
+type.
+
+For a receiver with struct type, the receiver is first copied, then each `member_initializer` is processed 
+the same way as an assignment to a field or property access of the result of the conversion. 
+Assignments are processed in lexical order.
+
+# Improvements on records
+
+## Allow `record class`
+
+The existing syntax for record types allows `record class` with the same meaning as `record`:
+
+```antlr
+record_declaration
+    : attributes? class_modifier* 'partial'? 'record' 'class'? identifier type_parameter_list?
+      parameter_list? record_base? type_parameter_constraints_clause* record_body
+    ;
+```
+
+## Allow user-defined positional members to be fields
+
+TODO
+
+See https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-10-05.md#changing-the-member-type-of-a-primary-constructor-parameter
+
+# References
+
+- https://github.com/dotnet/csharplang/edit/master/proposals/csharp-9.0/records.md 
+- https://github.com/dotnet/csharplang/blob/master/spec/structs.md
+- https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-05-04.md#structs-as-records
+- https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-22.md#struct-records
+- https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-09-30.md#struct-equality
+- https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-10-05.md#record-struct-primary-constructor-defaults
+- https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-10-07.md#record-struct-syntax
+
+# Open questions
+
+- should we disallow a user-defined constructor with a copy constructor signature?
+- confirm that we want to keep PrintMembers design (separate method returning `bool`)
+- confirm that we want to disallow members named "Clone".
+- why did we disallow unsafe types in instance fields in records? I assume we also want to disallow in record structs.
+- `with` on generics? (may affect the design for record structs)
```

Comment:
`record ref struct`?

Reply:
And if we do allow such a thing, can you use `ref` in the primary constructor to declare a `ref` field in conjunction with the addition of `ref` fields?

Reply:
Good question for sure. 
I'll capture that as an open issue.

---
In reply to: [552998468](https://github.com/dotnet/csharplang/pull/4307#discussion_r552998468) [](ancestors = 552998468)